Added State Function to JavalinRenderer

TareqK opened the pull request:
The Idea is to add a state function to the JavalinRenderer similar to the one in JavalinVue. This is needed, at least for me, because I need to do some filtering and customization on reports generated using templates depending on the user and a couple of other options.

TareqK commented:
@tipsy  just thought this might also be useful to others

tipsy commented:
Thanks @TareqK ! What people typically do for this is:

```kotlin
ctx.render("/template.ext", baseModel(ctx) + componentModel)
```

Since the `FileRenderer`  already has access to the `Context`, you could also create your own rendering implementation for setting a base model every time.

I sort of like taking the state-function concept from `JavalinVue`, but I'm not sure having implicit values in the template model is good. It is namedspace to `javalin`, but that could also be considered a downsiden (having to prefix each property).

What do you think about getting rid of the default properties and removing the namespace?

TareqK commented:
> What do you think about getting rid of the default properties and removing the namespace?

I think it would be a little inconsistent, since we inject javalin using $javalin in JavalinVue, i think it makes sense to keep it namespaced and with different properties for path/query/state . Also, we can be a little more sure that any data we inject unlikely to conflict with the javalin namespace. 

I think it really comes down to consistency and convenience;  I dont want to explicitly call and inject my state it into the model each time, or implement my own FileRenderer each time i want state injected.

TareqK commented:
@tipsy  so should we merge this or do you want me to change the namespace ?

tipsy commented:
It's named `$javalin` in `JavalinVue`, because that's the convention Vue uses for extending its prototype. For template models, on the other hand, the convention is to put everything in a `Map<String Any>`. I would rather these two plugins be consistent within their own domain than have a "Javalin-flavored" consistency cross-domain.

Before merging this, I would make the state function return a `Map<String, Object>`, and make the default implementation empty. You should also merge the component-specific state into the base-state, overriding any duplicate keys (and perhaps logging info?).

TareqK commented:
@tipsy  Ive gone ahead and refactored the state map and added overriding, though my Kotlin-Fu is limited so i just converted the model to a mutable map and modified that(but at least, the argument called stays immutable). Ive also added a ```state()``` utility method, similar to ```model()```, and ive added a per-render state to the ```Context``` class. Ive added tests for everything.

I also removed the namespace prefix, so you can access directly using ```queryParams``` and ```pathParams``` and ```state```. ive also made it so that array params are kept as arrays instead of being parsed to a string.

tipsy commented:
I think we need to take a step back here, adding a `state: Map<String, Any>` doesn't really give any added benefit to how templates are rendered (since they already have a `model: Map<String, Any>`). If I understand correctly what you want to achieve, all you need is your one function:

```kotlin
@JvmField
var stateFunction: (Context) -> Map<String, Any?> = { mapOf<String, Any>() }
```

Although, I would probably name is `baseModelFunction`, which would indicate that the `model` is merged on top of the result of the `baseModelFunction`.

The only other change required should be to this line:

```kotlin
return renderer.render(filePath, model, ctx)
```

So that it includes this new base model:

```kotlin
return renderer.render(filePath, baseModelFunction(ctx) + model, ctx)
```

The user can add query/path params themselves if they need them.

TareqK commented:
I assume since we are passing the CTX to the base function anyway, we can insert them from there right ? I'll patch it up tommorow - but do we still need a per-component state then ?

tipsy commented:
> I assume since we are passing the CTX to the base function anyway, we can insert them from there right ? 

Could you elaborate?

>  Do we still need a per-component state then ?

You'll still want per-template state (as in a model for each template). 
The base model will handle things like the current user, the current locale/language, if the website is using dark or light theme, etc. 
The template specific model with deal with things that are only relevant for a specific page, for example account details or a list of users being displayed.

tipsy commented:
This is out now ! https://javalin.io/news/2021/01/19/javalin-3.13.0-released.html
